pretyping/evarsolve.ml

```diff
@@ -1356,17 +1354,32 @@ let has_constrainable_free_vars env evd aliases force k ev (fv_rels,fv_ids,let_r
 
 exception EvarSolvedOnTheFly of evar_map * EConstr.constr
 
+let is_default = function
+| SList.Default (_, SList.Nil) | SList.Nil -> true
+| _ -> false
+
 (* Try to project evk1[argsv1] on evk2[argsv2], if [ev1] is a pattern on
    the common domain of definition *)
 let project_evar_on_evar force unify flags env evd aliases k2 pbty (evk1,argsv1 as ev1) (evk2,argsv2 as ev2) =
+  let evi1 = Evd.find evd evk1 in
+  let evi2 = Evd.find evd evk2 in
+  if Option.is_empty pbty && is_default argsv1 && is_default argsv2 &&
+    Option.is_empty evi1.evar_candidates && Option.is_empty evi2.evar_candidates &&
```

Comment:
It seems to me we could try to postpone the two `find` calls until we are 100% sure that none of the other conditions here are `false`. And even then we could delay the second lookup. Obviously that's ugly code to write in OCaml (and I won't even try to write it because I don't know how to do that in a nice way). But this MR does actually address a real problem and has performance benefits in some developments. It would be nice if we could get it to be as close to neutral as possible in the other developments.

The reason I think it might actually work is that the performance profile shows basically no regression in CPU instructions. That suggests to me that we might be suffering from cache misses or generally increased memory traffic. `find`, AFAICT, traverses a tree data structure so it could well be that that incurs additional cache misses.